### ISSUE TITLE
Make lazy initialization in plugins more robust

### ISSUE THREAD

**Issue by awaelchli** [PyTorchLightning/pytorch-lightning#7650]
## 🚀 Feature

We currently set some attributes like sync_batch_norm, num_nodes, devices etc. lazily in the accelerator connector, so the user does not have to provide them at instantiation. 

Example:

Trainer(gpus=4, plugins=DDPPlugin(find_unused_parameters=True))  # plugin may require gpus, num_nodes etc.

# AcceleratorConnector does this:
training_type_plugin.num_nodes = ...

This is fragile. Some attributes may have to be recomputed based on the order in which others are set.

### Pitch

Provide one single lazy init method that takes all arguments required. The plugin is responsible for making sure dependencies are resolved in one place: 

class DDPPlugin(...):

    def lazy_init(self, **kwargs):
         self.num_nodes = kwargs.get("num_nodes")
         self.num_processes = ...

**Comment by tchaton** [comment]
Enforcing a lazy mechanism for setting the parameters could reduce friction for customisation of the plugins as less logic would like within the AcceleratorConnector.

**Comment by carmocca** [comment]
So is this about having something similar to the dataclass `__post_init__`? Called by the Trainer? When would it be called exactly?

**Comment by awaelchli** [comment]
Yes exactly. It would be called once the accelerator connector has finished setup and instantiating all objects needed.

**Comment by joanovallesr** [comment]
Hi! I'd like to take a crack at implementing this. Since plugins have evolved into Strategies since this issue was opened, I want to confirm if the plan is still to implement a `lazy_init(**kwargs)` directly on the `DDPStrategy` class and refactor the `_AcceleratorConnector` to use it. If so, I have my environment set up and can start drafting a PR!
